lib/shelf_io.dart

```diff
@@ -31,6 +31,52 @@ import 'src/util.dart';
 
 export 'src/io_server.dart' show IOServer;
 
+/// Starts an [HttpServer] that listens on the specified [address] and
+/// [port] and sends requests to [handler].
+///
+/// If a [securityContext] is provided an HTTPS server will be started.
+///
+/// Serves requests in an error zone, using [onError] to handle errors, if
+/// provided. Uses a default error handler that prints error summary if
+/// [onError] is not provided.
+///
+/// See the documentation for [HttpServer.bind] and [HttpServer.bindSecure]
+/// for more details on [address], [port], [backlog], and [shared].
+Future<HttpServer> serveGuarded(
+  Handler handler,
+  address,
+  int port, {
+  SecurityContext? securityContext,
```

Comment:
If we are moving to a whole new API anyway we should spend some time investigating whether the design is the right one for current usage.

My gut is that wrapping the `HttpServer.bind` call and forwarding arguments like this isn't adding much value, and that the `serveRequests` API is readable enough if call sites need to inline the `HttpServer.bind`.

Reply:
> My gut is that wrapping the HttpServer.bind call and forwarding arguments like this isn't adding much value, and that the serveRequests API is readable enough if call sites need to inline the HttpServer.bind.

I am not really following - do you mean that `serve` and `serveGuarded` are not useful? I haven's seen much use of `serve` myself, maybe they both need to be removed. I only added a guarded version of the existing one.

Reply:
I think you are right and `serveXXX` are not very useful - normally users would have modifications to the starting the server code, be able to access the server (for stopping) and error handling in case it didn't start properly.

Reply:
Btw would you like me to just remove serve and serveRequests? Or do we need to deprecate them?

Reply:
I think we'll need to deprecate them. We might consider having a bit of time before we add the annotation so we can get some migrations out of the way ahead of CI turning red.

I've looked at this a little bit closer and I think I'm more convinced that we want to leave the `HttpServer` stuff to the clients. The caller already needs to be worried about calling `close()` on that server, so it makes sense to me to push the entirety of lifetime control for the server to the caller.

@kevmoo @jakemac53 - what do you think of trying something with an entirely different API?

```dart
extension ServeRequests on Handler {
  void serveRequests(Stream<HttpRequest> requests, {required void Function(Object, StackTrace) onError}) {
    runZonedGuarded(() {
      requests.forEach((request) => handleRequest(request, this));
    }, onError);
  }
}
```